File: py/ch01/E01C_reverse_complement.py
# Reverse Complement Problem:
# Find the reverse complement of a DNA string
#   Input: A DNA string Pattern
#   Output: Pattern, the reverse complement of Pattern

from dictionaries import complement_key, rna_complement_key


# Complexity: O(n)
def ReverseComplement(pattern, rna=False):
    result = ""

    for i in range(len(pattern) - 1, -1, -1):
        if rna:
            result += rna_complement_key.get(pattern[i])
        else:
            result += complement_key.get(pattern[i])

    return pattern, result


if __name__ == "__main__":
    _pattern = input("Pattern: ").upper()

    print(ReverseComplement(_pattern))

File: cpp/ch01/E01C_reverse_complement.cpp
//
// Created by Matej Hrnčiar on 21/07/2022.
//

#include "E01C_reverse_complement.h"

string ReverseComplement(const string &pattern) {
    string result = string(pattern.size(), 'X');
    int k = 0;

    for (int i = pattern.size() - 1; i >= 0; i--) {
        result[k] = complement_key.find(pattern[i])->second;
        k++;
    }

    return result;
}

/*
int main() {
    string _pattern;

    cout << "Pattern: ";
    cin >> _pattern;

    auto revComplement = ReverseComplement(_pattern);

    cout << "Original string: " << _pattern << endl << "Reverse complement: " << revComplement;
}
*/
